PR 2/5 - refactor(user): Implement BookmarkService (#878)

itsAidenDev opened the pull request:
This PR corresponds to Task [878](https://tree.taiga.io/project/luisvi-ita-challenges/task/878) and introduces the new BookmarkService layer, responsible for handling all logic related to user bookmarks.

The purpose of this implementation is to centralize bookmark-related operations into a dedicated service, improving modularity, scalability, and testability of the codebase.

It follows the work completed in task [877](https://tree.taiga.io/project/luisvi-ita-challenges/task/877) — Create"Bookmark" documents and repositories : preparation phase.

### 🔹 Changes Introduced

Created the new BookmarkService interface and BookmarkServiceImpl implementation class.

Implemented the main method:

`   public Mono<Set<UUID>> getUserBookmarks(String userId) 
`

This method, getUserBookmarks, retrieves a set of bookmarked challenge IDs for a given user. Here's a breakdown of its functionality:

- Takes a userId string parameter and validates it as a valid UUID
- Queries the user repository to find the user by their UUID
- Handles two error cases:
    ‣ Returns a BadUUIDException if the provided ID is invalid
    ‣ Returns a NotFoundException if no user is found with the given ID
- Returns a Mono<Set<UUID>> containing the user's bookmarked challenge IDs
- If the user has no bookmarks, returns an empty HashSet

This method is part of the reactive service layer, ensuring non-blocking operations when fetching user bookmarks from the database.

### 🔹 Advantages

- Promotes separation of concerns, isolating bookmarks logic from user management.
- Improves maintainability and testability of the backend architecture.
- Sets the foundation for future tasks involving the GET controller (Task [879](https://tree.taiga.io/project/luisvi-ita-challenges/task/879)) and integration tests (Task [880](https://tree.taiga.io/project/luisvi-ita-challenges/task/880)).
- Ensures a reactive, non-blocking design aligned with the new interactions architecture.

### ✅ Created / Updated Files

**Created**:

- BookmarkService.java
- BookmarkServiceImpl.java
- BookmarkServiceImplTest.java

**Updated:**

- UserService.java
- UserServiceImpl.java
- UserServiceImplTest.java
- UserController + UserControllerTest (temporary changes until the next PR, so the code can run properly)


### 💡 Notes

Files related to documents and repositories (e.g., BookmarkDocument, BookmarkResponseDto, BookmarkRepository, etc.) appear in this PR because PR 1/5 has not been merged into develop yet. These files are not part of the scope of this PR (2/5) and should be ignored during the review.

The service implementation directly uses the FavoriteRepository introduced in Task https://github.com/IT-Academy-BCN/ita-challenges-backend/pull/1040.

Future controller logic (Task [879](https://tree.taiga.io/project/luisvi-ita-challenges/task/879)) will consume this service to expose REST endpoints for retrieving user bookmarks.


✅ PR Author Self-Check

- [x]  I tested my changes locally and verified they work as expected.
- [x]  The PR has a clear and focused purpose.
- [x]  Title and description are filled in correctly.
- [x]  There are no leftover merge conflicts or unrelated changes from previous branches.
- [x]  All automated checks (like SonarCloud) have passed.
- [x]  I responded to all previous review comments and only resolved those I truly addressed.
- [x]  I ran SonarLint locally on the modified files and confirmed there are no warnings or errors

itsAidenDev committed:
Created bookmark docs and repo

itsAidenDev committed:
Created bookmark docs and repo

itsAidenDev committed:
Added more testing to BookmarkResponseDtoTest

itsAidenDev committed:
Added more testing to BookmarkResponseDtoTest

itsAidenDev committed:
Reversed changes in UserServiceImpl

itsAidenDev committed:
Reversed changes in UserServiceImpl

itsAidenDev committed:
Created new BookmarkService and BookmarkServiceImpl + testing

itsAidenDev committed:
Fix update in UserServiceImpl and UserServiceImplTest

zohra-b commented on itachallenge-user/src/main/java/com/itachallenge/user/controller/UserController.java:
```diff
@@ -37,10 +38,12 @@ public class UserController {
 
     private final UserService userService;
     private final IUserSolutionService userSolutionService;
+    private final BookmarkService bookmarkService;  // Add this line
```
remove this comment

zohra-b commented on itachallenge-user/src/main/java/com/itachallenge/user/service/UserServiceImpl.java:
```diff
@@ -105,18 +105,20 @@ private Mono<Boolean> addToFavorites(UUID userUuid, UUID challengeUuid) {
+        return userRepository.save(user)
+                .thenReturn(true)
+                .onErrorReturn(false);
     }
```
In the previoous Sprint, when implementing Favorites collection, it has been decided that the addFavorites method will no longer be filling the User Repobut that we will be filling the empty Favorites collection. 
I think you should be following the same implementation

zohra-b commented on itachallenge-user/src/main/java/com/itachallenge/user/service/UserServiceImpl.java:
```diff
@@ -129,18 +131,16 @@ private Mono<Boolean> deleteFromFavorites(UUID userId, UUID challengeUuid) {
+                .thenReturn(true)
+                .onErrorReturn(false);
     }
 
```
Same comment as for addToBookmarks. you should implement this method to use the Bookmark repository

zohra-b commented on itachallenge-user/src/main/java/com/itachallenge/userinteraction/service/bookmark/BookmarkServiceImpl.java:
```diff
@@ -0,0 +1,43 @@
+        } catch (IllegalArgumentException ex) {
+            return Mono.error(new BadUUIDException("Invalid ID format"));
+        }
+    }
```
you are not using the bookmark repo to get the favorites

itsAidenDev committed:
BookmarkRepository implemented in BookmarkServiceImpl (testing still needs fixing)

zohra-b committed:
fixed addToBookmarks method

zohra-b committed:
fixing tests in userServuceImpl

zohra-b committed:
tests fixed

zohra-b committed:
fixed sonarCloud issues

zohra-b committed:
Change JSON property name from 'uuid_favorite' to 'uuid_bookmark'

zohra-b committed:
Change JSON property name from 'uuid_favorite' to 'uuid_bookmark'

zohra-b committed:
Update JsonProperty value for uuid field in tests

zohra-b committed:
Update JsonProperty value for uuid field in tests

zohra-b committed:
Refactor BookmarkDocumentTest by removing redundant tests

Removed multiple tests related to equals, hashCode, and builder methods for BookmarkDocument.

zohra-b committed:
Refactor BookmarkDocumentTest by removing redundant tests

Removed multiple tests related to equals, hashCode, and builder methods for BookmarkDocument.

zohra-b committed:
Remove unused bookmarkDocument2 in tests

Removed unused bookmarkDocument2 initialization from tests.

zohra-b committed:
Remove unused bookmarkDocument2 in tests

Removed unused bookmarkDocument2 initialization from tests.

zohra-b committed:
Remove redundant equality and hashCode tests

Removed multiple tests related to equals, hashCode, and builder methods from BookmarkResponseDtoTest.

zohra-b committed:
Remove redundant equality and hashCode tests

Removed multiple tests related to equals, hashCode, and builder methods from BookmarkResponseDtoTest.

zohra-b committed:
Quitted @data to avoid Sonar request coverage

zohra-b committed:
Quitted @data to avoid Sonar request coverage

zohra-b committed:
Add missing closing brace in BookmarkDocumentTest

zohra-b committed:
Add missing closing brace in BookmarkDocumentTest

zohra-b committed:
fixed Sonar issue 

changed "exists ? // : // " to "exists == Boolean.TRUE ? // : //"

aperez1haad committed:
Revert "Replace data for getter/setter"

This reverts commit fb0a57fb77879f431e9e22e974d63005c40f4ff3.

aperez1haad committed:
Revert "Replace data for getter/setter"

This reverts commit fb0a57fb77879f431e9e22e974d63005c40f4ff3.

eduzavarce committed:
Refactor: Extract common fields to InteractionDocument and extend in BookmarkDocument and FavoriteDocument

eduzavarce committed:
Refactor: Extract common fields to InteractionDocument and extend in BookmarkDocument and FavoriteDocument

eduzavarce committed:
Remove unused @AllArgsConstructor import from BookmarkDocument

eduzavarce committed:
Remove unused @AllArgsConstructor import from BookmarkDocument

eduzavarce committed:
Remove redundant assertions and add missing test annotation in FavoriteDocumentTest

eduzavarce committed:
Remove redundant assertions and add missing test annotation in FavoriteDocumentTest

eduzavarce committed:
Consolidate and refactor equals, hashCode, and toString tests in BookmarkDocumentTest and FavoriteDocumentTest

eduzavarce committed:
Consolidate and refactor equals, hashCode, and toString tests in BookmarkDocumentTest and FavoriteDocumentTest

zohra-b committed:
updated changelog.md

zohra-b committed:
updated changelog.md

zohra-b committed:
Fix typo in CHANGELOG for InteractionDocument

zohra-b committed:
rebase with develop

zohra-b committed:
Merge branch 'refactor/877-create-bookmarks-docs-and-repo' into refactor/878-implement-bookmark-service

zohra-b committed:
made InteractionDocument class abstract

zohra-b committed:
quitted AllArgConstructor annotation

zohra-b committed:
turned InteractionDocument fields private (not protected) and turned to add AllArgConstructors for build was unsuccessful without

zohra-b committed:
quitted constructors in favoritedocument and in Bookmarkdocument to use SuperBuilder

zohra-b committed:
modified entire BookmarkDocumentTest class to maintain one only test

zohra-b committed:
quitted no arg / all arg test - coverage still 100

zohra-b committed:
simplified BookmarkResponseDto test : 2 tests remainings

zohra-b committed:
refactor BookmarkdocumentTest

zohra-b committed:
refactored FavoriteDocumentTest : one only test remaining

zohra-b committed:
refactored FavoriteResponseDtoTest : one test, coverage 100%

zohra-b committed:
Merge branch 'refactor/877-create-bookmarks-docs-and-repo' into refactor/878-implement-bookmark-service

zohra-b committed:
fixed a test in UserServiceImpltest

ivilarop committed:
test(userinteraction): simplify document tests and keep coverage

ivilarop committed:
test(userinteraction): simplify document tests and keep coverage

ivilarop committed:
test(userinteraction): simplify document tests and keep coverage

zohra-b committed:
Merge branch 'refactor/877-create-bookmarks-docs-and-repo' into refactor/878-implement-bookmark-service

ivilarop commented:
This pull request is not applicable because it contains many changes that have already been merged. We've opened a new pull request with the correct changes: https://github.com/IT-Academy-BCN/ita-challenges-backend/pull/1055
